helix-core/src/main/java/org/apache/helix/controller/stages/ManagementModeStage.java

```diff
@@ -157,14 +163,14 @@ private void recordClusterStatus(ClusterManagementMode mode, HelixDataAccessor a
       return;
```

Comment:
If it is a new state record, but it is fully frozen, we don't record the history? I guess this is not what we want.

Reply:
+1. I guess what we should do is:

if the current status is equal the status passed in, we skip it. Otherwise, we move forward to update it.

Reply:
@huizhilu pinged me and explained. This logic is to prevent misleading records if a certain participant gets any message after FROZEN. In such a case, without this block, we will record additional cluster freeze records.

I agree that we don't need a history record in this case. So the logic is good.

Could you please add an example the the comment so others won't be confused like me in the future? Thanks.

Reply:
I think this is still some room to improve. If it is "FROZEN" and get any message, the status is still "FROZEN" and we should not record any history. This will not cause duplicated record.

This should be fixed at logic level.

Reply:
Do you mean to pull the check logic outside to make the recording logic clean?
Sounds like a good plan to me.

Reply:
Thanks for the review, Guys.

This part is not a logic issue. Controller does not know the status has recorded **completed freeze mode**, until it reads the status znode. And the read logic is needed for updating the status. So the dedup logic can be and should be put in the record function. I treat them as a whole part of updating the status, including dedup.

To make the logic a bit cleaner, I've pulled the check logic out to another private method `shouldSkipRecording()`, which makes it easier to read and maintain.

Let me know if you have any questions or suggestions.